Use resumeAfter to resume changestream on 4.0

francoisferrand opened the pull request:
On 4.2, we should use `startAfter` parameter to resume the change stream.
However, this is not available on 4.0, so we must use `resumeAfter`
instead.

Changes:
- Use `resumeAfter` or `startAfter` depending on mongo version
- Use Change Steam wrapper in notification config manager to avoid duplicate logic

Issue: BB-386

francoisferrand committed:
Use resumeAfter to resume changestream on 4.0

On 4.2, we should use `startAfter` parameter to resume the change stream.
However, this is not available on 4.0, so we must use `resumeAfter`
instead.

Issue: BB-386

bert-e commented:
# Hello francoisferrand,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://bert-e.scality.net/scality/backbeat/doc/user).





*Status report is not available.*

williamlardier commented on extensions/oplogPopulator/OplogPopulator.js:
```diff
@@ -231,6 +233,7 @@ class OplogPopulator {
             pipeline: changeStreamPipeline,
             handler: this._handleChangeStreamChangeEvent.bind(this),
             throwOnError: false,
+            useStartAfter: !this._mongoVersion.startsWith("4.0.")
```
I am not sure this will work with the current BB CI, can we do this a bit more "semver"-like?

https://github.com/scality/backbeat/blob/development/8.6/.github/workflows/tests.yaml#L79

williamlardier replied:
Also, I am not sure the version is good enough. In case of upgrades, unless we explicitely set the featureVersion of mongodb, we might be in 4.2 but without `startAfter` support. Instead, you can do something like this:

```
db.adminCommand( 
    { 
        getParameter: 1, 
        featureCompatibilityVersion: 1 
    } 
)
```

it will return something like `{ "featureCompatibilityVersion" : { "version" : "4.2" }, "ok" : 1 }`, which will be better to test the compatibility

williamlardier reviewed (changes requested):
Some quick comments, plus some linter warnings that can be easily solved

Commit:
add util function for getting MongoDB version

Issue: BB-386

Commit:
use changeStream wrapper in NotificationConfigManager

remove duplicate logic for handling change streams by using the
already implemented change streams wrapper

Issue: BB-386

bert-e commented:
# Incorrect fix version



The `Fix Version/s` in issue BB-386 contains:


* *None*



Considering where you are trying to merge, I ignored possible hotfix versions and I expected to find:



* `8.6.16`


Please check the `Fix Version/s` of BB-386, or the target
branch of this pull request.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* 2 peers

francoisferrand commented on tests/functional/notification/NotificationConfigManager.js:
```diff
@@ -100,13 +104,11 @@ describe('NotificationConfigManager ::', () => {
             },
         };
-        manager._metastoreChangeStream.emit('change', changeStreamEvent);
+        manager._changeStreamWrapper._changeStream.emit('change', changeStreamEvent);
```
it would have been clearer to keep the existing name (_metastoreChangeStream), even if the type changes : it better conveys the intention that we listen to metadata (vs just knowing this is a change stream, where we need "outside" knowledge to know what we use it)

```suggestion
        manager._metastoreChangeStream._changeStream.emit('change', changeStreamEvent);
```

francoisferrand commented:
LGTM (but I can't approve since I was the one creating the PR initially)

francoisferrand commented:
/approve

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/8.6`



The following branches will **NOT** be impacted:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/8.6`



The following branches have **NOT** changed:

* `development/7.10`
* `development/7.4`
* `development/7.70`
* `development/8.4`
* `development/8.5`



Please check the status of the associated issue BB-386.

Goodbye francoisferrand.
